File: py/wkmb.py
import math
from mathlib import cumdist, sqL2, clustercenter

class WKM:
  """
  Warped K-Means: an algorithm to cluster sequentially-distributed data.
  Luis A. Leiva and Enrique Vidal. Information Sciences, 2013.
  See http://dx.doi.org/10.1016/j.ins.2013.02.042
  Implemented by Luis A. Leiva.
  Dual licensed under the MIT and GPL licenses.
  Project page: http://personales.upv.es/luileito/wkm/
  """
  
  def __init__(self, samples, numclusters, threshold=0.0):
    """
    Instantiate the class.
    Threshold is the cluster population that will be tested in each iteration.
    """
    self.maxiter = 100
    # User input
    self.samples     = samples
    self.numclusters = int(numclusters)
    self.threshold   = float(threshold)
    # Assume that all samples have the same dimensions
    self.dimensions = len(self.samples[0])
    # Sanitize user input
    if self.numclusters < 1:
      self.numclusters = 1
    elif self.numclusters > len(samples):
      self.numclusters = len(samples)
    if threshold > 1:
      self.threshold = 1.0
    elif threshold < 0:
      self.threshold = 0.0
    self.reset()


  def reset(self):
    """Clean up."""
    self.initialized  = False
    # System ouput
    self.boundaries   = [0] * self.numclusters
    self.clusters     = [0] * self.numclusters
    self.centroids    = [0] * self.numclusters
    self.localenergy  = [0] * self.numclusters
    self.totalenergy  = 0.0 # FIXME: actually it can be computed as array sum of localenergy
    self.iterations   = 0
    self.numtransfers = 0
    self.cost         = 0


  def init(self, method=None):
    """Initialization method."""
    self.reset()
    N, M = len(self.samples), self.numclusters
    # Silly checks
    if self.numclusters <= 1: # single partition
      self.boundaries = [0]
      return  
    elif self.numclusters >= N: # singleton clusters
      self.boundaries = [i for i in range(N)]
      return
    # Finally check user-defined method
    if method == None:
      self.initdefault(N,M)
    else:
      m = method.lower() 
      if m == "ts":
        self.TS(N,M)
      elif m == "eq":
        self.resample(N,M)
      else: 
        self.initdefault(N,M)


  def initdefault(self, N, M):
    """
    Default boundaries initialization. 
    Will use TS only if 2M <= N (~ Nyquist).
    Other methods may be implemented, as long as they process samples in a sequential fashion.
    """
    self.boundaries = []
    if N/float(M) < 2:
      self.resample(N,M)
    else:
      self.TS(N,M)
    self.resample(N,M)
    
    
  def TS(self, N, M):
    """
    Initialize boundaries by trace segmentation (non-linear allocation). 
    This is the pre-set initialization mode.
    """
    self.boundaries = []
    Lcum, LN = cumdist(self.samples)
    incr, i = LN / float(M), 0
    for j in range(1,M+1):
      fact = (j - 1)*incr
      while fact > Lcum[i] or i in self.boundaries:
        i += 1
      self.boundaries.append(i) 
    self.initialized = True
    
    
  def resample(self, N, M):
    """Allocate N points into M boundaries in a linear fashion (see t8 APP)."""
    self.boundaries = []
    b = -1
    for i in range(N):
      q = math.floor( (i+1)*M/(N+1.0) )
      if q > b:
        b = q
        self.boundaries.append(i)
    self.initialized = True
    

  def getPartition(self):
    """Assign points to a cluster in a sequential fashion."""
    for j in range(self.numclusters):
      self.clusters[j] = self.getClusterSamples(j)
      assert len(self.clusters[j]) > 0, "Empty cluster %d" % j


  def getClusterSamples(self, index):
    """Retrieve points by cluster index."""
    l = self.boundaries[index+1] if (index+1 < self.numclusters) else len(self.samples)
    return self.samples[self.boundaries[index]:l]


  def setPartition(self, partition):
    """Specify a sequential cluster configuration."""
    self.boundaries, self.clusters = [0], []
    for j, points in enumerate(partition):
      assert len(points) > 0, "Empty cluster %d" % j
      self.clusters[j] = points
      if j > 0:
        self.boundaries.append(len(points)-1)


  def computeEnergies(self):
    """Compute the energy of all clusters from scratch."""
    self.totalenergy = 0.0
    for j in range(self.numclusters):
      points = self.clusters[j]
      assert len(points) > 0, "Empty cluster %d" % j
      self.centroids[j] = clustercenter(points)
      energy = 0.0
      for pt in points:
        energy += sqL2(pt, self.centroids[j])
      self.localenergy[j] = energy
      self.totalenergy += energy


  def incrementalMeans(self, sample, j, bestcluster, n, m):
    """Recompute cluster means as a result of reallocating a sample."""
    newj = [0.0] * self.dimensions
    newbest = newj[:]
    for d in range(self.dimensions):
      newbest[d] = self.centroids[bestcluster][d] + (sample[d] - self.centroids[bestcluster][d]) / (m+1.0)
      newj[d] = self.centroids[j][d] - (sample[d] - self.centroids[j][d]) / (n-1.0)
    self.centroids[bestcluster] = newbest
    self.centroids[j] = newj
    

  def cluster(self, partition=None):
    """Perform sequential clustering."""
    if not self.initialized:
      self.init()
    if partition:
      self.setPartition(partition)
    else:
      self.getPartition()
    self.computeEnergies()
    # Silly check
    if self.numclusters < 2:
      return
    # Reallocate boundaries
    while True:
      transfers = False # no transfers yet
      for j in range(self.numclusters):
        points = self.clusters[j]
        n = len(points)
        if n < 2: continue
        # Otherwise explore 1st half
        if j > 0: 
          for i in range(0, int(math.floor(n/2.0 * (1 - self.threshold))) + 1):
            sample = points[i]
            bestcluster = j-1
            m = len(self.clusters[bestcluster])
            n = len(self.clusters[j])
            J1 = (m / (m + 1.0)) * sqL2(sample, self.centroids[bestcluster])
            J2 = (n / (n - 1.0)) * sqL2(sample, self.centroids[j])
            delta = J1 - J2
            self.cost += 1
            if delta < 0 and n > 1:
              transfers = True
              self.numtransfers += 1
              self.boundaries[j] += 1
              # Recompute means incrementally
              self.incrementalMeans(sample,j,bestcluster,n,m)
              # Recompute local energies
              self.localenergy[bestcluster] += J1
              self.localenergy[j] -= J2
              # Overall energy simply decreases
              self.totalenergy += delta
              self.getPartition()
            else: break
        points = self.clusters[j]
        n = len(points)
        if n < 2: continue
        # Otherwise explore 2nd half
        if j + 1 < self.numclusters:
          for i in range(n-1, int(math.floor(n/2.0 * (1 + self.threshold))) - 2, -1):
            sample = points[i]
            bestcluster = j+1
            m = len(self.clusters[bestcluster])
            n = len(self.clusters[j])
            J1 = (m / (m + 1.0)) * sqL2(sample, self.centroids[bestcluster])
            J2 = (n / (n - 1.0)) * sqL2(sample, self.centroids[j])
            delta = J1 - J2
            self.cost += 1
            if delta < 0 and n > 1:
              transfers = True
              self.numtransfers += 1
              self.boundaries[bestcluster] -= 1
              # Recompute means incrementally
              self.incrementalMeans(sample,j,bestcluster,n,m)
              # Recompute local energies
              self.localenergy[bestcluster] += J1
              self.localenergy[j] -= J2
              # Overall energy simply decreases
              self.totalenergy += delta
              self.getPartition()
            else: break
      # Update
      self.iterations += 1
      if not transfers or self.iterations == self.maxiter: break
    # Finally, recompute energies from scratch when algorithm converges, to avoid rounding errors
    self.computeEnergies()

File: js/wkmb.js
(function(module, global) {

  var math   = require("./mathlib")
    , assert = require('assert')
    ;

  /**
   * Warped K-Means: an algorithm to cluster sequentially-distributed data.
   * Luis A. Leiva and Enrique Vidal. Information Sciences, 2013.
   * See http://dx.doi.org/10.1016/j.ins.2013.02.042
   * @author Luis A. Leiva
   * @license Dual licensed under the MIT and LGPL licenses.
   * @see Project page: http://personales.upv.es/luileito/wkm/
   */
  var WKM = function(samples, numclusters, threshold) {
    // Instantiates the class.
    // Threshold is the cluster population that will be tested in each iteration.
    this.maxiter = 100;
    // User input
    this.samples     = samples;
    this.numclusters = parseInt(numclusters);
    this.threshold   = parseFloat(threshold) || 0;
    // Assume that all samples have the same dimensions
    this.dimensions = samples[0].length;
    // Sanitize user input
    if (this.numclusters < 1) {
      this.numclusters = 1;
    } else if (this.numclusters > samples.length) {
      this.numclusters = samples.length;
    }
    if (threshold > 1) {
      this.threshold = 1;
    } else if (threshold < 0) {
      this.threshold = 0;
    }
    this.reset();
  };

  /**
   * Clean up.
   * @return void
   */
  WKM.prototype.reset = function() {
    this.initialized  = false;
    // System ouput
    this.boundaries   = [];
    this.clusters     = [];
    this.centroids    = [];
    this.localenergy  = [];
    this.totalenergy  = 0;
    this.iterations   = 0;
    this.numtransfers = 0;
    this.cost         = 0;
  };

  /**
   * Initialization, optionally using a given method.
   * @param method String "ts" or "eq"
   * @return void
   */
  WKM.prototype.init = function(method) {
    this.reset();
    var N = this.samples.length, M = this.numclusters;
    // Silly checks
    if (this.numclusters <= 1) { // single partition
      this.boundaries = [0];
      return;
    } else if (this.numclusters >= N) { // singleton clusters
      for (var i = 0; i < M; i++) {
        this.boundaries[i] = i;
      }
      return;
    }
    // Finally check user-defined method
    if (typeof method === 'undefined') {
      this.initdefault(N,M);
    } else {
      method = method.toLowerCase();
      if (method == "ts") {
        this.TS(N,M);
      } else if (method == "eq") {
        this.resample(N,M);
      } else {
        this.initdefault(N,M);
      }
    }
  };

  /**
   * Default boundaries initialization. 
   * Will use TS only if 2M <= N (~ Nyquist).
   * Other methods may be implemented, as long as they process samples in a sequential fashion.
   * @param N Number Number of samples
   * @param M Number Number of clusters
   * @return void
   */
  WKM.prototype.initdefault = function(N, M) {
    this.boundaries = [];
    if (N/M < 2) {
      this.resample(N,M);
    } else {
      this.TS(N,M);
    }
    this.resample(N,M)
  };

  /**
   * Initialize boundaries by trace segmentation (non-linear allocation). 
   * @param N Number Number of samples
   * @param M Number Number of clusters
   * @return void
   */
  WKM.prototype.TS = function(N, M) {
    this.boundaries = [];
    var ret = math.cumdist(this.samples),
        Lcum = ret.values, 
        LN   = ret.length,
        incr = LN/M, 
        i = 0;
    for (var j = 1; j < M+1; j++) {
      var fact = (j - 1)*incr;
      while (fact > Lcum[i] || this.boundaries.indexOf(i) > -1) {
        i++;
      }
      this.boundaries.push(i);
    }
    this.initialized = true;
  };
  
  /**
   * Allocate N points into M boundaries in a linear fashion.
   * @param N Number Number of samples
   * @param M Number Number of clusters
   * @return void
   */
  WKM.prototype.resample = function(N, M) {
    this.boundaries = [];
    var b = -1;
    for (var i = 0; i < N; i++) {
      var q = Math.floor( (i+1)*M/(N+1) );
      if (q > b) {
        b = q;
        this.boundaries.push(i);
      }
    }
    this.initialized = true;
  };      

  /** 
   * Assign points to a cluster in a sequential fashion.
   * @return void
   */
  WKM.prototype.getPartition = function() {
    for (var j = 0; j < this.numclusters; j++) {
      this.clusters[j] = this.getClusterSamples(j);
      //assert(this.clusters[j].length > 0, "Empty cluster " + j);
    }
  };

  /** 
   * Retrieve points by cluster index.
   * @param index Number Cluster index
   * @return Cluster samples
   */
  WKM.prototype.getClusterSamples = function(index) {
    var l = index+1 < this.numclusters ? this.boundaries[index+1] : this.samples.length;
    return this.samples.slice(this.boundaries[index], l);
  };

  /** 
   * Specify a sequential cluster configuration.
   * @param partition Array Cluster
   * @return void
   */
  WKM.prototype.setPartition = function(partition) {
    this.boundaries = [0];
    this.clusters = [];
    for (var j = 0; j < partition.length; j++) {
      var points = partition[j];
      //assert(points.length > 0, "Empty cluster " + j);
      this.clusters[j] = points.slice();
      if (j > 0) {
        this.boundaries.push(points.length - 1);
      }
    }
  };

  /** 
   * Compute the energy of all clusters from scratch.
   * @return void
   */
  WKM.prototype.computeEnergies = function() {
    this.totalenergy = 0;
    for (var j = 0; j < this.numclusters; j++) {
      var points = this.clusters[j];
      //assert(points.length > 0, "Empty cluster " + j);
      this.centroids[j] = math.clustercenter(points);
      var energy = 0;
      for (var i = 0; i < points.length; i++) {
        energy += math.sqL2(points[i], this.centroids[j]);
      }
      this.localenergy[j] = energy;
      this.totalenergy += energy;
    }
  };

  /** 
   * Recompute cluster means as a result of reallocating a sample to a better cluster.
   * @return void
   */
  WKM.prototype.incrementalMeans = function(sample, j, b, n, m) {
    var d, newj = [];
    var newb = newj.slice();
    for (d = 0; d < this.dimensions; d++) {
      newb[d] = this.centroids[b][d] + (sample[d] - this.centroids[b][d]) / (m + 1);
      newj[d] = this.centroids[j][d] - (sample[d] - this.centroids[j][d]) / (n - 1);
    }
    this.centroids[b] = newb;
    this.centroids[j] = newj;
  };
  
  /** 
   * Perform sequential clustering.
   * @return void
   */
  WKM.prototype.cluster = function(partition) {
    if (!this.initialized) this.init();
    if (partition) this.setPartition(partition);
    else this.getPartition();
    this.computeEnergies();
    // Silly check
    if (this.numclusters < 2) return;
    // Reallocate boundaries
    var i, j, b, c, n, m, J1, J2, delta, p, transfers;
    while (true) {
      transfers = false; // no transfers yet
      for (j = 0; j < this.numclusters; j++) {
        if (j > 0) {
          c = this.clusters[j].slice();
          n = c.length;
          // Reallocate backward 1st half
          for (i = 0; i <= Math.floor(n/2 * (1 - this.threshold)) + 1; i++) {
            p = c[i];
            b = j - 1;
            m = this.clusters[b].length;
            n = this.clusters[j].length;
            if (n < 2) break;
            J1 = (m / (m + 1)) * math.sqL2(p, this.centroids[b]);
            J2 = (n / (n - 1)) * math.sqL2(p, this.centroids[j]);
            delta = J1 - J2;
            this.cost++;
            if (delta < 0) {
              transfers = true;
              this.numtransfers++;
              this.boundaries[j]++;
              this.incrementalMeans(p,j,b,n,m);
              this.localenergy[b] += J1;
              this.localenergy[j] -= J2;
              this.totalenergy += delta;
              this.getPartition();
            } else break;
          }
        }
        if (j + 1 < this.numclusters) {
          c = this.clusters[j].slice();
          n = c.length;
          // Reallocate forward 2nd half
          for (i = n - 1; i >= Math.floor(n/2 * (1 + this.threshold)) - 2; i--) {
            p = c[i];
            b = j + 1;
            m = this.clusters[b].length;
            n = this.clusters[j].length;
            if (n < 2) break;
            J1 = (m / (m + 1)) * math.sqL2(p, this.centroids[b]);
            J2 = (n / (n - 1)) * math.sqL2(p, this.centroids[j]);
            delta = J1 - J2;
            this.cost++;
            if (delta < 0) {
              transfers = true;
              this.numtransfers++;
              this.boundaries[b]--;
              this.incrementalMeans(p,j,b,n,m);
              this.localenergy[b] += J1;
              this.localenergy[j] -= J2;
              this.totalenergy += delta;
              this.getPartition();
            } else break;
          }
        }
      }
      this.iterations++;
      if (!transfers || this.iterations == this.maxiter) break;
    }
    // Finally, recompute energies from scratch when algorithm converges, to avoid rounding errors
    this.computeEnergies();  
  };

  module.exports = WKM;

})('object' === typeof module ? module : {}, this);
